text/0322-cdk-pipelines-updated-api.md

```diff
@@ -0,0 +1,455 @@
+> better, and provide a more streamlined user-facing API,
+> both while staying backwards compatible.
+
+## Working Backwards
+
+The final API for CDK Pipelines is finally here. Compared to the previous API:
+
+- CDK applications could not be deployed in waves. They can now, as we automatically establish runorders between
+  actions.
+- You could not build pipelines that used multiple sources to feed into the synth, or that had a custom build step. You can now.
+- Large CDK applications (> 25 stacks) could not previously be deployed because they would exceed the 50 actions per
+  stage limit that CodePipeline has. With the new implementation, the actions will automatically spread over multiple
+  stages, even if you deploy multiple applications in parallel.
+- For simple use cases, you no longer need to manage CodePipeline Artifacts: artifact management is implicit
+  as objects that produce artifacts can be used to reference them.
+
+
+### Examples
+
+The following:
+
+```ts
+import { SecretValue } from '@aws-cdk/core';
+import * as codepipeline from '@aws-cdk/aws-codepipeline';
+import * as codepipeline_actions from '@aws-cdk/aws-codepipeline-actions';
+import * as cdkp from '@aws-cdk/pipelines';
+
+const sourceArtifact = new codepipeline.Artifact();
+const cloudAssemblyArtifact = new codepipeline.Artifact('CloudAsm');
+const integTestArtifact = new codepipeline.Artifact('IntegTests');
+
+const pipeline = new cdkp.CdkPipeline(this, 'Pipeline', {
+  cloudAssemblyArtifact,
+
+  // Where the source can be found
+  sourceAction: new codepipeline_actions.GitHubSourceAction({
+    actionName: 'GitHub',
+    output: sourceArtifact,
+    oauthToken: SecretValue.secretsManager('github-token'),
+    owner: 'OWNER'
+    repo: 'REPO',
+    trigger: codepipeline_actions.GitHubTrigger.POLL,
+  }),
+
+  // How it will be built
+  synthAction: cdkp.SimpleSynthAction.standardNpmSynth({
+    sourceArtifact,
+    cloudAssemblyArtifact,
+    projectName: 'MyServicePipeline-synth',
+    additionalArtifacts: [
+      {
+        directory: 'test',
+        artifact: integTestArtifact,
+      },
+    ],
+  }),
+});
+
+const stage = new MyStage(this, 'PreProd', {
+  env: { account: '12345', region: 'us-east-1' },
+});
+stage.addActions(
+  new cdkp.ShellScriptAction({
+    commands: ['node ./integ-tests'],
+    additionalArtifacts: [integTestArtifact],
+  }),
+);
+```
+
+Becomes:
+
+```ts
+import * as cdkp from '@aws-cdk/pipelines';
+
+const pipeline = new cdkp.Pipeline(this, 'Pipeline', {
+  build: new cdkp.CdkBuild({
+    input: cdkp.CodePipelineSource.gitHub('OWNER/REPO'),
+    commands: ['npm ci', 'npm run build'],
+    additionalOutputs: {
+      tests: cdkp.AdditionalOutput.fromDirectory('test'),
+    }
+  }),
+  backend: new cdkp.AwsCodePipelineBackend(),
+});
+
+const stage = new MyStage(this, 'PreProd', {
+  env: { account: '12345', region: 'us-east-1' },
+});
+pipeline.addCdkStage(stage, {
```

Comment:
Is it possible to also support:

```ts
const stage = pipeline.addStage(..);
stage.addAfterAction(new cdkp.ShellScriptAction(...));
```

Should we call it "post" instead of "after"?

Reply:
> Is it possible to also support:

Yes, for sure. Since everything is just deeply mutable object soup, every attribute can be done either at creation or by calling adders.

> Should we call it "post" instead of "after"?

If we want to sound fancy, sure :)